kompari/src/imgdiff.rs

```diff
@@ -0,0 +1,102 @@
+// Copyright 2024 the Kompari Authors
+// SPDX-License-Identifier: Apache-2.0 OR MIT
+
+use image::{Pixel, Rgba};
+use std::fmt::{Debug, Formatter};
+
+use crate::Image;
+
+pub enum ImageDifference {
+    None,
+    SizeMismatch {
+        left_size: (u32, u32),
+        right_size: (u32, u32),
+    },
+    Content {
+        n_different_pixels: u64,
+        distance_sum: u64,
+        diff_image: Image,
+    },
+}
+
+impl Debug for ImageDifference {
+    fn fmt(&self, f: &mut Formatter<'_>) -> std::fmt::Result {
+        match self {
+            Self::None => write!(f, "Difference::None"),
+            Self::SizeMismatch {
+                left_size,
+                right_size,
+            } => write!(
+                f,
+                "Difference::SizeMismatch({:?}, {:?})",
+                left_size, right_size
+            ),
+            Self::Content {
+                n_different_pixels, ..
+            } => f
+                .debug_struct("Difference::Content")
+                .field("n_different_pixels", n_different_pixels)
+                .finish(),
+        }
+    }
+}
+
+/// Find differences between two images
+pub fn compare_images(left: &Image, right: &Image) -> ImageDifference {
+    if left.width() != right.width() || left.height() != right.height() {
+        return ImageDifference::SizeMismatch {
+            left_size: (left.width(), left.height()),
+            right_size: (right.width(), right.height()),
+        };
+    }
+
+    let n_different_pixels: u64 = left
+        .pixels()
+        .zip(right.pixels())
+        .map(|(pl, pr)| if pl == pr { 0 } else { 1 })
+        .sum();
+
+    if n_different_pixels == 0 {
+        return ImageDifference::None;
+    }
+
+    let mut distance_sum: u64 = 0;
+
+    let diff_image_data: Vec<u8> = left
+        .pixels()
+        .zip(right.pixels())
+        .flat_map(|(&p1, &p2)| {
+            let total_distance = pixel_distance(p1, p2);
+            distance_sum += total_distance;
+            if total_distance > 0 {
+                // Highlight the differences atop a semi-transparent view of the shared parts.
+                p2.0
```

Comment:
This is my fault. I think the original green/red logic should be restored. I do think that showing the previous "background" (the `else` block) is useful.

We could make this a follow-up, but this is a new logic.

Reply:
I have returned red/green diff, but I like your idea. My next follow-up will be some UI parts that will allow to switch diff visualization.